src/libexpr/primops/filterPath.cc

```diff
@@ -0,0 +1,174 @@
+#include "primops.hh"
+
+namespace nix {
+
+struct FilteringInputAccessor : InputAccessor
+{
+    EvalState & state;
+    PosIdx pos;
+    ref<InputAccessor> next;
+    CanonPath prefix;
+    Value * filterFun;
+
+    std::map<CanonPath, bool> cache;
+
+    FilteringInputAccessor(EvalState & state, PosIdx pos, const SourcePath & src, Value * filterFun)
+        : state(state)
+        , pos(pos)
+        , next(src.accessor)
+        , prefix(src.path)
+        , filterFun(filterFun)
+    {
+    }
+
+    std::string readFile(const CanonPath & path) override
+    {
+        checkAccess(path);
+        return next->readFile(prefix + path);
+    }
+
+    bool pathExists(const CanonPath & path) override
+    {
+        checkAccess(path);
+        return next->pathExists(prefix + path);
+    }
+
+    Stat lstat(const CanonPath & path) override
+    {
+        checkAccess(path);
+        return next->lstat(prefix + path);
+    }
+
+    DirEntries readDirectory(const CanonPath & path) override
+    {
+        checkAccess(path);
+        DirEntries entries;
+        for (auto & entry : next->readDirectory(prefix + path)) {
+            if (isAllowed(path + entry.first))
+                entries.insert(std::move(entry));
+        }
+        return entries;
+    }
+
+    std::string readLink(const CanonPath & path) override
+    {
+        checkAccess(path);
+        return next->readLink(prefix + path);
+    }
+
+    void checkAccess(const CanonPath & path)
+    {
+        if (!isAllowed(path))
+            throw Error("access to path '%s' has been filtered out", showPath(path));
```

Comment:
```suggestion
        if (!isAllowed(path))
            throw Error("access to path '%s' has been filtered out by filterPath", showPath(path));
```

Also this will lead to surprising errors like `access to path 'defualt.nix' has been filtered out by filterPath`, leading someone to debug their filter function rather than fix the typo.

I've tried to improve the error message in a suggestion here, but to get enough information, we need a `pathExistsUnfiltered` function.